### DISCUSSION TITLE
Trouble with Pi Zero W and the second version of Centaur

### DISCUSSION THREAD

**Discussion by dgtlsoul** [EdNekebno/DGTCentaurMods#131] [Q&A] [answered]
Thanks for the amazing work.  I was very impressed by how the card-setup-tool and firstboot.sh automate everything.  All seemed to have gone well, but the board did not start up at the end as expected.  Here is what I have:

- DGTCentaurMods 1.1.0
- Pi Zero W (not 2)
- Second version of Centaur (WFT0290CZ10LW)

I left the board on for about 30 minutes to finish firstboot.sh.  I can do SSH to verify that the board is running, but I cannot access its web interface thru http://[ip]:5000.  The e-ink display stays at the welcome screen “Let’s Play” and the only button I can use is the power switch. 

Is using Pi Zero W instead of Pi Zero 2W a potential cause of the problem?  I bought a new Centaur because I wanted to experiment with your code so much.  Any suggestion would be greatly appreciated.  Here is my [centaur_install.log](https://github.com/EdNekebno/DGTCentaurMods/files/8218808/centaur_install.log).

**Comment by wormstein** [comment]
Hi,
I had a look on your install log (using the mobile) and i certainly see that DGTCentaurMods finished installing.

One thing for sure i can tell you is that the mod works with pi zero w. A bit slow on install, but works.

Second, mod is not showing the DGT Centaur logo with Let's play text. That is the original software and there is no way for it to start on boot.

Something else must have happened. Maybe you mistaken the cards?

Things you could try:
- check for DGTCM in /opt
-`sudo systemctl start DGTCentaurMods.service` it should auto start anyway...
- start all over again. Your install log is showing mod did installed.

Bonus: the last card-setup-tool in `master` implements a new feature that will show progress on the ePaper on self install. This is not released yet but still works with 1.1.0
Note, it takes few minutes for the OS to prepare and init the screen

Let me know how it goes.

Cheers

**Comment by wormstein** [comment, marked as answer]
Likely the logo from orig software remained on eink as it did no refresh... Nevermind this.

Your problem could be a known bug where a serial comm death lock happens.

When this happens, ssh to your board and do a `sudo systemctl restart DGTCentaurMods.service`

We knew about it before 1.1.0 and tried to fix. Looks like it's still around sometimes... 
We did more fixes to address this in our next release. It should be out this weekend.

-A

**Reply by dgtlsoul** [reply]
Thanks for your prompt reply.  The logo was indeed just residual from the original software.  I do have the right card in place.

I tried `sudo systemctl restart DGTCentaurMods.service` but that didn’t do the trick.  I started over with `card-setup-tool` in master, but still ran into the same problem.  I wonder why the web interface would not start at port 5000.  Would a serial comm death lock affect the web interface?

![Screen Shot 2022-03-09 at 5 44 49 PM](https://user-images.githubusercontent.com/1624928/157572206-c6291d2d-c599-44e8-b9f5-0b7d9daf0c8b.png)

I also wonder if my soldering skills may be the cause.  Does this look like a poor job to you? 😅

![image1](https://user-images.githubusercontent.com/1624928/157572552-0310a3eb-5dd7-4da6-b984-6ae783dab913.png)

**Reply by wormstein** [reply]
Sorry my fault, the tool in master is not yet the one mentioned :(

**Reply by wormstein** [reply]
Hard to tell on soldering....  SPI might not work so no output to the screen then. 

The mod service is running fine. Web interface is on other service though. Try this to debug:

```
sudo systemctl stop centaurmods-web.service
cd /opt/DGTCentaurMods/web
flask run
```

See the output.

**Reply by wormstein** [reply]
Ok, silly me. Just double checked.

So card setup tool in master IS the new one.

If nothing was shown on screen during setup then you likely have a hardware issue with the soldering. SPI is not working. Screen uses SPI...

For the web GUI, try the step in my last post. 
Web GUI is showing a realtime image of the screen, so that feature could be the cause the web interface fails to start.

Cheers

**Reply by wormstein** [reply]
Also, debug the mod service:

```
sudo stop DGTCentaurMods.service
cd /opt/DGTCentaurMods
export PYTHONPATH=/opt
python game/menu.py
```

Check if any exception while running. Let's see...

I incline to say this is h/w issue at this point.

**Reply by dgtlsoul** [reply]
Following your instruction, I found out `python game/menu.py` stopped at `Detecting board adress`.  I looked into `board.py` and saw it got stuck at `while len(resp) < 4:` since `rest = ser.read(1000)` never got anything.  

If I bypass that `while` loop to use the default `addr1` and `addr2`, it then gets stuck at `Attempting to clear serial`.  At this point, I could get Flask running, although it cannot detect any piece on the chess board.

This looks like a pretty clear indication of a hardware issue.  I will redo the soldering or try a different Pi Zero W.  I will report back if I have any progress.

Thank you for being so responsive and supportive!  🙏

**Reply by dgtlsoul** [reply]
Re-soldering the pins fixed the issue!  Everything runs beautifully now.   

I did suspect the issue to be hardware-related in the beginning, but not being able to see the web interface made me want to look into the software further.  If that `while` loop could exit with an error after enough attempts and pass it to the web interface, it may help people with poor soldering skills like myself find the root cause more quickly :)

Nonetheless, I'm in awe of your dedication and support!  Thank you for sticking with me!

**Reply by wormstein** [reply]
Glad it's working for you. Don't forget to enable updates :)
The board woun't function if no address is detected. It might with the default `addr1` and `addr2` but is not certain.

The problem on web interface is that it imports the entire board module so it will execute the address detection again. Two options here: either cherry pick what to import or refactor the board module and put addr detection in a function and call it.
The project started with modules writen as blocks that will execute on import insted of being more object oriented. I started slowly to prototype some classes but refactoring everything is HUGE work and planning. Ah, and more contributors... :))

**Reply by wormstein** [reply]
Ah, we're on discord too in case you need support.

**Reply by dgtlsoul** [reply]
I'm personally very interested in this project.  I will have some free time from April and will see if there is anything I can contribute.  Keep up the awesome work!
